apollos-church-api/src/data/ContentItem.js

```diff
@@ -236,8 +236,25 @@ class dataSource extends ContentItem.dataSource {
         resources.map(
           async ({ attributeValues: { linkName, linkUrl, file } }) => {
             let url = '';
-            if (linkUrl.value) {
-              url = ApollosConfig.ROCK.URL + linkUrl.value;
+
+            function isValidHttpUrl(string) {
+              let checkUrl;
+
+              try {
+                checkUrl = new URL(string);
+              } catch (_) {
+                return false;
+              }
+
+              return (
+                checkUrl.protocol === 'http:' || checkUrl.protocol === 'https:'
+              );
+            }
+
+            if (isValidHttpUrl(linkUrl.value)) {
```

Comment:
I think you might be over complicating this a little bit. From a code perspective, the only _REAL_ issue you have here is that this function definition is inline of a mapper function - which means this function is getting redefined for every type this map is ran - so like if you have 3 `resources`, the function gets redefined 3 different types. IT's a performance issue. You could define this function at the top/root of this file and you'd get better performance.

However, you're already doing `new URL` below (lines 255 and 257) regardless if the URL is valid or not, you're just adding `ApollosConfig.ROCK.URL` as the second argument if it's not. You might be able to simplify this to something like....

```
if (linkUrl.value) {
  url = new URL(linkUrl.value, !linkUrl.value.startsWith('http') ? ApollosConfig.ROCK.URL : undefined);
} else if (file.value) {
```

And skip the `isValidHttpUrl` stuff all together. Does this make sense? Basically if the string doesn't start with `http`, then you know it's not a valid url, and the Rock URL needs to be added, otherwise the `new URL` bit will validate the URL anyways. This way, you're only calling `new URL` once.

Reply:
Yeah, that makes sense.